### PR TITLE
🌱 feat: control plane deletion checks

### LINKED ISSUE #2176: Error deleting cluster with managed node pool
/kind bug
/area provider/eks

**What steps did you take and what happened:**
Running the upgrade test, when tearing down the cluster in the cleanup there is an error:

```
error deleting EKS cluster for EKS control plane eks-upgrade-3ffvrg/cluster-a65xnm-control-plane: unable to delete EKS cluster: failed to request delete of eks cluster eks-upgrade-3ffvrg_cluster-a65xnm-control-plane: ResourceInUseException: Cluster has nodegroups attached
```

The test logs are [here](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/2175/pull-cluster-api-provider-aws-e2e-eks/1343269651437588480/artifacts/clusters/bootstrap/controllers/capa-eks-control-plane-controller-manager/capa-eks-control-plane-controller-manager-79bfb6965b-8smg6/manager.log)

**What did you expect to happen:**
I expect the tear down to be done without error. So any managed nodegroups need to be deleted first.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api-provider-aws version: 0.6.3

### LINKED ISSUE #2179: [e2e]  EKS e2e tests are failing due on reconciling aws-auth configmap
/kind bug
/area provider/eks

**What steps did you take and what happened:**
When running the EKS e2e tests the aws-iam-authenticator config map reconciliation is failing:

```bash
E1229 22:51:49.003887       1 controller.go:257] controller-runtime/controller "msg"="Reconciler error" "error"="failed to reconcile aws-iam-authenticator config for AWSManagedControlPlane eks-nodes-hxozyw/cluster-aoqabq-control-plane: getting client for remote cluster: Get https://EF1A2963DEFAB7D55742686FCD350535.gr7.us-west-2.eks.amazonaws.com/api?timeout=32s: dial tcp 34.210.95.236:443: i/o timeout" "controller"="awsmanagedcontrolplane" "name"="cluster-aoqabq-control-plane" "namespace"="eks-nodes-hxozyw"
```

**What did you expect to happen:**
The EKS e2e tests should pass this and the config map should be reconciled

**Anything else you would like to add:**
The control plane manager logs with an example are [here](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/2175/pull-cluster-api-provider-aws-e2e-eks/1344046080525864960/artifacts/clusters/bootstrap/controllers/capa-eks-control-plane-controller-manager/capa-eks-control-plane-controller-manager-6c9995688-ntpw8/manager.log)

**Environment:**

- Cluster-api-provider-aws version: 0.6.3

### PR DESCRIPTION
**Pull request by richardcase**
**What this PR does / why we need it**:
When investigating #2179  a number of issues have been discovered and this PR fixes them. When the control plane is deleted a check is made to see if there are any existing dependencies...if there are then the request is requeued until the dependencies no longer exist. 

This is required to enable the upgrade tests but it doesn't fix #2179 (but additional logging will help).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2176

### CODE REVIEW

File: controlplane/eks/controllers/awsmanagedcontrolplane_controller.go [hunk 1]
```diff
@@ -326,3 +351,41 @@ func (r *AWSManagedControlPlaneReconciler) managedClusterToManagedControlPlane(o
...
+	dependencies := 0
+
+	machines := &infrav1.AWSMachineList{}
+	if err := r.Client.List(ctx, machines, listOptions...); err != nil {
+		return dependencies, fmt.Errorf("failed to list machines for cluster %s/%s: %w", namespace, clusterName, err)
+	}
+	r.Log.V(2).Info("tested for AWSMachine dependencies", "count", len(machines.Items))
+	dependencies += len(machines.Items)
+
+	if feature.Gates.Enabled(feature.MachinePool) {
```
Comment (michaelbeaumont) [inline controlplane/eks/controllers/awsmanagedcontrolplane_controller.go:374]: I'm wondering if it's worth perhaps just using the API to check for both nodegroups _and_ fargate profiles (which also prevent deletion)? Or are dependencies created outside of CAPI out of scope here?

Comment (richardcase) [reply controlplane/eks/controllers/awsmanagedcontrolplane_controller.go:374]: I did think about that but my thinking is that the CAPA CRD types are the dependency which in turn own the AWS resources and I didn't want the control plane to be aware of the implementation details of each of the other controllers.

Comment (michaelbeaumont) [reply controlplane/eks/controllers/awsmanagedcontrolplane_controller.go:374]: Yeah that's totally valid 👍

File: bootstrap/eks/config/rbac/leader_election_role_binding.yaml [hunk 2]
```diff
@@ -8,5 +8,5 @@ roleRef:
   name: leader-election-role
 subjects:
 - kind: ServiceAccount
-  name: default
+  name: controller-manager
```
Comment (michaelbeaumont) [inline bootstrap/eks/config/rbac/leader_election_role_binding.yaml:11]: I ran into this too, should it be a separate PR?

**Review by michaelbeaumont — APPROVED** [review]

File: bootstrap/eks/config/rbac/leader_election_role_binding.yaml (hunk 2 above)
Comment (richardcase) [reply bootstrap/eks/config/rbac/leader_election_role_binding.yaml:11]: Yes it should, i have created #2183

Comment (richardcase) [reply bootstrap/eks/config/rbac/leader_election_role_binding.yaml]: Separate PR created and this has been removed from this change

**Comment by richardcase** [comment]
@michaelbeaumont - i've created a separate PR for the leader election changes.

File: controlplane/eks/controllers/awsmanagedcontrolplane_controller.go [hunk 3]
```diff
@@ -238,11 +248,21 @@ func (r *AWSManagedControlPlaneReconciler) reconcileNormal(ctx context.Context,
...
 
-func (r *AWSManagedControlPlaneReconciler) reconcileDelete(_ context.Context, managedScope *scope.ManagedControlPlaneScope) (_ ctrl.Result, reterr error) {
+func (r *AWSManagedControlPlaneReconciler) reconcileDelete(ctx context.Context, managedScope *scope.ManagedControlPlaneScope) (_ ctrl.Result, reterr error) {
 	managedScope.Info("Reconciling AWSManagedClusterPlane delete")
 
 	controlPlane := managedScope.ControlPlane
 
+	numDependencies, err := r.dependencyCount(ctx, managedScope)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("error getting controlplane dependencies %s/%s: %w", controlPlane.Namespace, controlPlane.Name, err)
```
Comment (randomvariable) [inline controlplane/eks/controllers/awsmanagedcontrolplane_controller.go:258]:
```suggestion
                r.Log.Error(err, "error getting controlplane dependencies")
		return reconcile.Result{}, err)
```
Usual thing about getting the line of code the error comes from before it returns to controller runtime.

**Review by randomvariable — COMMENTED** [review]
nit around logging errors, but otherwise lgtm

File: controlplane/eks/controllers/awsmanagedcontrolplane_controller.go (hunk 3 above)
Comment (richardcase) [reply controlplane/eks/controllers/awsmanagedcontrolplane_controller.go]: Thanks @randomvariable, changed

**Pull request merged**